RecoTracker/TkDetLayers/src/GeometricSearchTrackerBuilder.cc

```diff
@@ -46,149 +42,120 @@ GeometricSearchTracker *GeometricSearchTrackerBuilder::build(const GeometricDet
-
-      for (vector<const GeometricDet *>::const_iterator it2 = theTOBGeometricDetLayers.begin();
-           it2 != theTOBGeometricDetLayers.end();
-           it2++) {
-        theTOBLayers.push_back(aPhase2OTBarrelLayerBuilder.build(*it2, theGeomDetGeometry, useBrothers));
+    if (theGeomDetLayer->type() == GeometricDet::OTPhase2Barrel) {
+      vector<const GeometricDet *> theTOBGeometricDetLayers = theGeomDetLayer->components();
+      for (const GeometricDet *thisGeomDet : theTOBGeometricDetLayers) {
+        theTOBLayers.push_back(aPhase2OTBarrelLayerBuilder.build(thisGeomDet, theGeomDetGeometry, useBrothers));
       }
     }
 
-    if ((*it)->type() == GeometricDet::PixelEndCap) {
-      vector<const GeometricDet *> thePxlFwdGeometricDetLayers = (*it)->components();
-      for (vector<const GeometricDet *>::const_iterator it2 = thePxlFwdGeometricDetLayers.begin();
-           it2 != thePxlFwdGeometricDetLayers.end();
-           it2++) {
-        if ((*it2)->positionBounds().z() < 0)
-          theNegPxlFwdLayers.push_back(aPixelForwardLayerBuilder.build(*it2, theGeomDetGeometry));
-        if ((*it2)->positionBounds().z() > 0)
-          thePosPxlFwdLayers.push_back(aPixelForwardLayerBuilder.build(*it2, theGeomDetGeometry));
+    if (theGeomDetLayer->type() == GeometricDet::PixelEndCap) {
+      vector<const GeometricDet *> thePxlFwdGeometricDetLayers = theGeomDetLayer->components();
+      for (const GeometricDet *thisGeomDet : thePxlFwdGeometricDetLayers) {
+        if ((thisGeomDet)->positionBounds().z() < 0)
+          theNegPxlFwdLayers.push_back(aPixelForwardLayerBuilder.build(thisGeomDet, theGeomDetGeometry));
+        else if ((thisGeomDet)->positionBounds().z() > 0)
+          thePosPxlFwdLayers.push_back(aPixelForwardLayerBuilder.build(thisGeomDet, theGeomDetGeometry));
       }
     }
 
-    if ((*it)->type() == GeometricDet::PixelPhase1EndCap) {
-      vector<const GeometricDet *> thePxlFwdGeometricDetLayers = (*it)->components();
-      for (vector<const GeometricDet *>::const_iterator it2 = thePxlFwdGeometricDetLayers.begin();
-           it2 != thePxlFwdGeometricDetLayers.end();
-           it2++) {
-        if ((*it2)->positionBounds().z() < 0)
-          theNegPxlFwdLayers.push_back(aPhase1PixelForwardLayerBuilder.build(*it2, theGeomDetGeometry));
-        if ((*it2)->positionBounds().z() > 0)
-          thePosPxlFwdLayers.push_back(aPhase1PixelForwardLayerBuilder.build(*it2, theGeomDetGeometry));
+    if (theGeomDetLayer->type() == GeometricDet::PixelPhase1EndCap) {
+      vector<const GeometricDet *> thePxlFwdGeometricDetLayers = theGeomDetLayer->components();
+      for (const GeometricDet *thisGeomDet : thePxlFwdGeometricDetLayers) {
+        if ((thisGeomDet)->positionBounds().z() < 0)
+          theNegPxlFwdLayers.push_back(aPhase1PixelForwardLayerBuilder.build(thisGeomDet, theGeomDetGeometry));
+        else if ((thisGeomDet)->positionBounds().z() > 0)
```

Comment:
else without the test for z>0?

Reply:
> else without the test for z>0?

Thank you for your comments, Vincenzo
I don't understand this one though...

Reply:
it can be either z<0 or z>0 (z==0 is not possible)
so  if is not z<0 MUST BE z>0 no need to test

Reply:
there are many other occurrency below

Reply:
> there are many other occurrencies below

Thank you @VinInn 

I think I fixed all them, but this one:
https://github.com/cms-sw/cmssw/pull/33420/files#diff-0b80efba96a5198928106c0248e4976130ec9316fde648c6738d77951bf2f0b3R119
because it explicitly admitted a further `else` statement already in the original code, which triggers a `LogError` in its turn